pandas/core/groupby.py

```diff
@@ -1487,19 +1489,32 @@ def aggregate(self, values, how, axis=0):
                 raise NotImplementedError
             out_shape = (self.ngroups,) + values.shape[1:]
 
-        if is_numeric_dtype(values.dtype):
-            values = com.ensure_float(values)
-            is_numeric = True
-            out_dtype = 'f%d' % values.dtype.itemsize
-        else:
-            is_numeric = issubclass(values.dtype.type, (np.datetime64,
-                                                        np.timedelta64))
+        if _is_datetime_or_timedelta_dtype(values.dtype):
+            values = values.view('int64')
+        elif is_bool_dtype(values.dtype):
+            values = _algos.ensure_float64(values)
+
+        is_numeric = is_numeric_dtype(values.dtype)
+        upsample = self.ngroups > values.size
+
+        if is_numeric and upsample:
```

Comment:
the problem is that you are filling an integer series with np.nan, which doesn't make sense (and it is converted to -int64_max which is the same as ``iNaT``). So soln is if its datetime/timedelta is to fill with ``iNaT`` rather than ``np.nan` (which is essentially the NaN for ints)`. Then after the computation, you can convert to float and replace with ``np.nan``, as you should be returning an integer series.

You can do this generically by seeing if the result dtype is int, if it is, then simply do:
```
(Pdb) x= result.astype('float64')
(Pdb) x[x==tslib.iNaT] = np.nan
```
It think converting to float of the results is ok, though if there are NO missing values, then you can leave as ints.